xarray/tests/test_units.py

```diff
@@ -527,12 +550,17 @@ def test_align_dataset(fill_value, unit, variant, error, dtype):
         strip_units(ds1), strip_units(convert_units(ds2, units_a)), **stripped_kwargs
     )
     expected_a = attach_units(expected_a, units_a)
-    expected_b = convert_units(attach_units(expected_b, units_a), units_b)
+    if isinstance(array2, Quantity):
+        expected_b = convert_units(attach_units(expected_b, units_a), units_b)
+    else:
+        expected_b = attach_units(expected_b, units_b)
 
     actual_a, actual_b = func(ds1, ds2)
 
-    assert_equal_with_units(expected_a, actual_a)
-    assert_equal_with_units(expected_b, actual_b)
+    assert_allclose(expected_a, actual_a)
+    assert extract_units(expected_a) == extract_units(actual_a)
+    assert_allclose(expected_b, actual_b)
+    assert extract_units(expected_b) == extract_units(actual_b)
```

Comment:
I'm using a combination of `assert_allclose` + `extract_units` and `assert_identical` now. The output of `assert_allclose` is not really ideal because only the values (no labels and no units) are printed. The output of `assert_identical` is also not ideal because the `repr` of quantities prints the units after the values. Since long strings get truncated, the unit is almost always not visible.

Reply:
> assert_allclose is not really ideal because only the values (no labels and no units) are printed. 

Yeah this needs to be updated to pretty-print the diffs like assert_equal.

> The output of assert_identical is also not ideal because the repr of quantities prints the units after the values. Since long strings get truncated, the unit is almost always not visible.

There was some talk about adding units explicitly to the repr. Seems like this would be a good thing to do early and will help you debug these tests more easily. Can you open an issue or PR for that?

Reply:
I think I can do the formatting for `assert_allclose`, but I don't know where to start for the repr for units. Maybe I should open a new issue for that?

Reply:
>  Maybe I should open a new issue for that?

let's talk about it on #2773